tests/tests_type_list.cpp

```diff
@@ -0,0 +1,207 @@
+		static_assert(std::is_same_v<tl::nth_t<t2, 0>, int>);
+		static_assert(std::is_same_v<tl::nth_t<t2, 1>, double>);
+		static_assert(std::is_same_v<tl::opt_t<tl::nth<t2, 2>>, tl::nullopt>);
+	}
+
+	TEST_CASE("concat") {
+		static_assert(std::is_same_v<tl::concat_t<empty_t, empty_t>, empty_t>);
+
+		using t1 = tl::type_list<int>;
+
+		static_assert(std::is_same_v<tl::concat_t<empty_t, t1>, t1>);
+		static_assert(std::is_same_v<tl::concat_t<t1, empty_t>, t1>);
+
+		using t2 = tl::type_list<int, double>;
+
+		static_assert(std::is_same_v<tl::concat_t<t1, t2>, tl::type_list<int, int, double>>);
+		static_assert(std::is_same_v<tl::concat_t<t2, t1>, tl::type_list<int, double, int>>);
+	}
+
+	TEST_CASE("apply") {
+		using t1 = tl::type_list<int>;
+		static_assert(std::is_same_v<tl::apply_t<t1, std::vector>, std::vector<int>>);
+
+		using t2 = tl::type_list<int, std::pmr::polymorphic_allocator<int>>;
+		static_assert(std::is_same_v<tl::apply_t<t2, std::vector>, std::vector<int, std::pmr::polymorphic_allocator<int>>>);
+	}
+
+	TEST_CASE("transform") {
+		constexpr auto func = []<typename T>(std::type_identity<T>) {
+			return std::add_const<T>{};
+		};
+
+		static_assert(std::is_same_v<tl::transform_t<empty_t, func>, empty_t>);
+
+		using t = tl::type_list<int, double, char>;
+		static_assert(std::is_same_v<tl::transform_t<t, func>, tl::type_list<int const, double const, char const>>);
+	}
+
+	TEST_CASE("filter") {
+		constexpr auto pred = []<typename T>(std::type_identity<T>) {
+			return std::is_const_v<T>;
+		};
+
+		static_assert(std::is_same_v<tl::filter_t<empty_t, pred>, empty_t>);
+
+		using t1 = tl::type_list<int, int const, double, char const>;
+		static_assert(std::is_same_v<tl::filter_t<t1, pred>, tl::type_list<int const, char const>>);
+	}
+
+	TEST_CASE("generate") {
+		static_assert(std::is_same_v<tl::generate_t<0, [] { }>, empty_t>);
+
+		using t = tl::generate_t<3, []<size_t idx>(std::integral_constant<size_t, idx>) {
+			if constexpr (idx == 0) {
+				return std::type_identity<int>{};
+			} else {
+				return std::type_identity<double>{};
+			}
+		}>;
+
+		static_assert(std::is_same_v<t, tl::type_list<int, double, double>>);
+	}
+
+	TEST_CASE("find_if") {
+		constexpr auto pred1 = []<typename T>(std::type_identity<T>) {
+			return std::is_const_v<T>;
+		};
+
+		constexpr auto pred2 = []<typename T>(std::type_identity<T>) {
+			return std::is_same_v<T, int>;
+		};
+
+		constexpr auto pred3 = []<typename T>(std::type_identity<T>) {
+			return std::is_same_v<T, double>;
+		};
+
+		constexpr auto pred4 = []<typename T>(std::type_identity<T>) {
+			return false;
+		};
+
+		static_assert(std::is_same_v<tl::opt_t<tl::find_if<empty_t, pred1>>, tl::nullopt>);
+
+		using t1 = tl::type_list<int, int const, double>;
+		static_assert(std::is_same_v<tl::find_if_t<t1, pred1>, int const>);
+		static_assert(std::is_same_v<tl::find_if_t<t1, pred2>, int>);
+		static_assert(std::is_same_v<tl::find_if_t<t1, pred3>, double>);
```

Comment:
do we need a test case for checking whether one of `pred1`, `pred2`, or `pred3` fail?